project-koku/koku
# [COST-5227] - Add auto large provider logic and XL polling

## Jira Ticket

[COST-5227](https://issues.redhat.com/browse/COST-5227)

## Description

Add a new boolean field to providers in order to indicate whether that provider should be classed as XL (large) or not. This includes changes to the manifest creation to determine and set whether the provider is large based on report count. The default is larger than 75 files though this is configurable.

Additionally adding a new report downloader schedule specifically for XL (large) customer polling. Currently these providers get collected with our standard schedule but that also means its possible that all XL providers could be scheduled at the same time, which can causes overloaded queues and alerts. This PR now schedules XL (large providers) on there own cadence and batch size giving us more control over load. 

Updated masu download endpoint with param to trigger large providers.

## Testing

1. Checkout Branch
2. Restart Koku
3. Load data 
4. See everything process like normal.
5. Either set `XL_REPORT_COUNT` in your env to 0 (or something lower than your next provider generates) or just hard code a value in place of `if num_of_files > settings.XL_REPORT_COUNT:`
6. Create a new provider with enough report files to class it as large.
7. Check the logs for `Provider set as large for XL processing.`
8. Update the `POLLING_TIMER` like report count so you can trigger downloads (set it to 1 second or something) `polling_delta = datetime.now(tz=settings.UTC) - timedelta(seconds=1) #settings.POLLING_TIMER)`
9. hit download endpoint with `?large_providers=True`
10. See in the logs that only the large provider is collected.
11. Do the same without the large flag
12. See that all providers are collected apart from the large provider.

Notes:
If you want to test multiple large providers to test polling batch size you can run the following sql to set other providers to large, just sub the provider uuid with your cloud provider `update api_provider set large = True where uuid = '4f991071-31d6-4c43-b4e3-4f6049e5c390';`

Both the XL batch size and the schedule cadence is customisable. 

## Release Notes

```markdown
* [COST-5227](https://issues.redhat.com/browse/COST-5227) Add automatic large providers and XL specific polling
```
